walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
+   
+    // Optional attendes.
+    List<Event> optionalBookedEvents = findBookedEvents(events, request.getOptionalAttendees());
+
+    // All attendees - Mandatory + Optional.
+    List<Event> allBookedEvents = getAllUniqueEvents(mandatoryBookedEvents, optionalBookedEvents);
+    allBookedEvents = sortEventsByStartTime(allBookedEvents);
+
+    // No possible meeting if request is longer than a day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+    // No events/no attendess means meeting can be anytime.
+    if(events.isEmpty() || (mandatoryBookedEvents.isEmpty() && optionalBookedEvents.isEmpty())) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+
+    // If mandatory attendees are available all day with no booked events.
+    // Try and find available times including optional event restrictions and return unless.
+    // no time that works with optional attendess works just return mandatory availabilities (All day).
+    if(mandatoryBookedEvents.isEmpty()){
+      Collection<TimeRange> allAvailable = findPotentialTimes(optionalBookedEvents, request);
+      
+      if(allAvailable.isEmpty()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+      return allAvailable;
+    }
+
+    // Try finding time with all attendees, if none can be found try finding time with
+    // only mandatory attendees.
+    Collection<TimeRange> times = findPotentialTimes(allBookedEvents, request);
+    if(times.isEmpty()){
+      return findPotentialTimes(mandatoryBookedEvents, request);
+    }
+    return times;
   }
+
+  /**
+   * Finds all the events that attendees in the request are booked for.
+   * @param events - all possible events.
+   * @param attendees - from the meeting request.
+   * @return events that request attendees are booked for.
+   */
+  public List<Event> findBookedEvents(Collection<Event> events, Collection<String> attendees) {
+    // List of existing bookedEvents for attendees in meeting request.
+    List<Event> bookedEvents = new ArrayList<>();
+
+    // Find all events that attendees in the request are already booked for.
+    for(Event event : events){
+      for (String attendee : attendees){
+        if(event.getAttendees().contains(attendee) && !bookedEvents.contains(event)){
+          bookedEvents.add(event);
+        }
+      }
+    }
+    return bookedEvents;
+  } 
+  /**
+   * Finds all potential availabilities in between bookedEvents.
+   * @param bookedEvents - timerange already booked out/unavailable.
+   * @param request - meetin request to add in a potential new timerange.
+   * @return list of timeranges for availability.
+   */
+  public Collection<TimeRange> findPotentialTimes(List<Event> bookedEvents, MeetingRequest request){
+    // List of potential free timeranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
+    
+    Event prevEvent = bookedEvents.get(0); 
```

Comment:
What is bookedEvents is empty?

Reply:
That's a good point. As the function is currently used there's a check to make sure it's not before the function is called, but that may not always be the case. Someone might come along and reuse the function, or refactor the code that calls it in such a way that the list might be empty.

A function should always handle any 'legal' input correctly. You can declare certain inputs 'illegal' by stating preconditions in the function comment, e.g. "Preconditions: 'bookedEvents' must be non-empty", but you should always check all those preconditions in the functions and throw an informative error if they aren't met.

For collections the caller will generally expect a function to be able to handle an empty list (a common case) so you should usually handle this without raising an error, i.e. by adding an early exit that returns a sensible value immediately if the input list is empty.

Reply:
Done.